dask/dataframe/core.py

```diff
@@ -1873,39 +1951,94 @@ def map_partitions(func, columns, *args, **kwargs):
         return Scalar(merge(dask, *[arg.dask for arg in args]), name)
 
     args = _maybe_from_pandas(args)
-
-    if columns is no_default:
-        columns = None
-
     from .multi import _maybe_align_partitions
     args = _maybe_align_partitions(args)
     dfs = [df for df in args if isinstance(df, _Frame)]
 
-    return_type = _get_return_type(dfs[0], columns)
+    if columns is no_default:
+        # pass no_default as much, because it updates internal cache
+        try:
+            columns = _emulate(func, *args, **kwargs)
+        except Exception:
+            # user function may fail
+            columns = None
+    metadata = columns
+
+    return_type = _get_return_type(dfs[0], metadata)
+
     dsk = {}
     for i in range(dfs[0].npartitions):
         values = [(arg._name, i if isinstance(arg, _Frame) else 0)
                   if isinstance(arg, (_Frame, Scalar)) else arg for arg in args]
-        dsk[(name, i)] = (_rename, columns, (apply, func, (tuple, values),
-                                                          kwargs))
+        values = (apply, func, (tuple, values), kwargs)
+        dsk[(name, i)] = (_rename, metadata, values)
 
     dasks = [arg.dask for arg in args if isinstance(arg, (_Frame, Scalar))]
-
-    return return_type(merge(dsk, *dasks), name, columns, args[0].divisions)
+    return return_type(merge(dsk, *dasks), name, metadata, args[0].divisions)
 
 
 def _rename(columns, df):
-    """ Rename columns in dataframe or series """
+    """
+    Destructively rename columns of pd.DataFrame or name of pd.Series.
+    Not for dd.DataFrame or dd.Series.
```

Comment:
Putting impure operations with the dask graph concerns me.  Other unknown computations may rely on those values and we can't rely on a particular ordering.

Is it possible to rely only on pure operations like `pd.DataFrame.rename`?